app/Utils/DatabaseCleanupUtils.php

```diff
@@ -117,6 +118,8 @@ public static function removeBuild($buildid) : void
             $buildids[] = intval($b);
         }
 
+        $builds = Build::whereIn('id', $buildids)->get();
```

Comment:
Eager loading should be used wherever possible to prevent iteratively querying lazy relations.

```suggestion
        $builds = Build::whereIn('id', $buildids)
            ->with('configure')
            ->withCount('configure.builds')
            ->get();
```

Reply:
Oops, I overlooked the possibility of a configure being shared by only builds being deleted.  That'll make things a bit more complicated.  A similar concept, except with [constrained eager loading](https://laravel.com/docs/11.x/eloquent-relationships#constraining-eager-loads) should work though.

Reply:
When you get a chance, please take a look at the latest commit I pushed to this branch. Does eager loading still make sense if we're going to be using `$builds` to check multiple relationships?

Reply:
I played around with this locally and it seemed like `whereBelongsTo` requires a strict `belongsTo` (not Many) relationship. Instead I ended up using `whereHas` and `whereDoesntHave`, and this seems to work well.

I also took your advice and manually inspected the query log and it looks reasonable to me:

```
delete from `configure` where exists (
  select * from `build`
  inner join `build2configure` on `build2configure`.`buildid` = `build`.`id`
  where `configure`.`id` = `build2configure`.`configureid` and `id` in (?))
and not exists (
  select * from `build`
  inner join `build2configure` on `build2configure`.`buildid` = `build`.`id`
  where `configure`.`id` = `build2configure`.`configureid` and `id` not in (?))
```

Reply:
@zackgalbreath That query seems more reasonable to me.  I think it would be worthwhile to check the query plan on a production database before merging though.  If you could post the `EXPLAIN` results for that query (you might need to run `SET @@explain_format=TREE;` first if it's not already set), that would be great.

Reply:
If I'm reading this correctly, the new query plan seems slightly better for potentially deleting a shared configure row when removing a single build. `cost=0.754` vs. `cost=0.947`.

___

New query:
```
delete from `configure` where exists (
  select * from `build`
  inner join `build2configure` on `build2configure`.`buildid` = `build`.`id`
  where `configure`.`id` = `build2configure`.`configureid` and `id` in (10089131))
and not exists (
  select * from `build`
  inner join `build2configure` on `build2configure`.`buildid` = `build`.`id`
  where `configure`.`id` = `build2configure`.`configureid` and `id` not in (10089131));
```

New plan:
```
| -> Delete from configure (immediate)  (cost=0.754 rows=2)
    -> Nested loop antijoin  (cost=0.754 rows=2)
        -> Rows fetched before execution  (cost=0..0 rows=1)
        -> Nested loop inner join  (cost=1.45 rows=2)
            -> Filter: ((build2configure.configureid = '1210940') and (build2configure.buildid <> 10089131))  (cost=0.754 rows=2)
                -> Covering index range scan on build2configure using configureid over (configureid = 1210940 AND buildid < 10089131) OR (configureid = 1210940 AND 10089131 < buildid)  (cost=0.754 rows=2)
            -> Single-row covering index lookup on build using PRIMARY (id=build2configure.buildid)  (cost=0.6 rows=1)
 |
```

___

Old query:
```
SELECT a.configureid
FROM build2configure AS a
LEFT JOIN build2configure AS b ON (
    a.configureid=b.configureid
    AND b.buildid NOT IN (10089131)
)
WHERE a.buildid IN (10089131)
GROUP BY a.configureid
HAVING count(b.configureid)=0;
```

Old plan:
```
| -> Filter: (count(b.configureid) = 0)  (cost=0.947 rows=1.41)
    -> Group aggregate: count(b.configureid)  (cost=0.947 rows=1.41)
        -> Nested loop left join  (cost=0.747 rows=2)
            -> Rows fetched before execution  (cost=0..0 rows=1)
            -> Filter: ((b.configureid = '1210940') and (b.buildid <> 10089131))  (cost=0.747 rows=2)
                -> Covering index range scan on b using configureid over (configureid = 1210940 AND buildid < 10089131) OR (configureid = 1210940 AND 10089131 < buildid)  (cost=0.747 rows=2)
 |
```

Reply:
The total cost difference isn't particularly meaningful here, particularly since it looks like it's probably anticipating that it will use cached rows, but it's good to see that it's in the same rough ballpark as the original query.  The new query plan does show what we need to see though: every lookup is done efficiently, and the database doesn't expect to use an excessive amount of rows for any component of the plan (i.e., memory usage will be low).  Importantly, the entire process is done atomically and we avoid needing to execute multiple queries.